dataloader/dataloader_impl.hpp

```diff
@@ -0,0 +1,142 @@
+/**
+ * @file dataloader_impl.hpp
+ * @author Eugene Freyman
+ * @author Mehul Kumar Nirala.
+ * @author Zoltan Somogyi
+ * @author Kartik Dutt
+ * 
+ * Implementation of DataLoader.
+ * 
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MODELS_DATALOADER_IMPL_HPP
+#define MODELS_DATALOADER_IMPL_HPP
+
+#include "dataloader.hpp"
+
+using namespace mlpack;
+
+template<
+  typename DataSetX,
+  typename DataSetY,
+  class ScalerType
+>DataLoader<
+    DataSetX, DataSetY, ScalerType
+>::DataLoader()
+{
+  // Nothing to do here.
+}
+
+template<
+  typename DataSetX,
+  typename DataSetY,
+  class ScalerType
+>DataLoader<
+    DataSetX, DataSetY, ScalerType
+>::DataLoader(const std::string &dataset,
+              const bool shuffle,
+              const double ratio,
+              const bool useScaler,
+              const std::vector<std::string> augmentation,
+              const double augmentationProbability)
+{
+  if (dataset == "mnist")
```

Comment:
Maybe it would be worthwhile to hold some struct that matches names of well-known datasets?  i.e.

```
// short name => details.  Maybe this can all be initialized in one line, I haven't tried.
const std::map<std::string, DatasetDetails> datasetMap;
datasetMap["mnist"] = Datasets::MNIST();
...
```

I think I have my syntax wrong there, but you get the idea I assume. :)

Then, in this function you could just check if `dataset` is in the keys of `datasets` and use that `DatasetDetails`, thus making the code below not specific to MNIST. :+1: